Generate HELM charts

Sgitario opened the pull request:
See helm-example to see how it would look like.
The solution is to add a new helm module to generate HELM objects.

Fix https://github.com/dekorateio/dekorate/issues/673
Fix https://github.com/dekorateio/dekorate/issues/844

Sgitario commented:
At the moment, the `values.yaml` is generated using all the Dekorate configuration coming from properties and annotations, but it's not used.

Following this comment: https://github.com/dekorateio/dekorate/issues/673#issuecomment-1000838521, the first approach will be to use the above `values.yaml` and try to map the Dekorate configuration to the Kubernetes configuration (this is what the decorators do).
If not suitable solution is found, we will implement it using option A (jkube) ++ B (set of known properties)

Sgitario commented:
JKube implementation can be found here:
- Module: https://github.com/eclipse/jkube/tree/master/jkube-kit/resource/helm
- Entry point: https://github.com/eclipse/jkube/blob/082e18a4f38e51e183ffa2a05321bcb5a5fd2ba0/jkube-kit/resource/helm/src/main/java/org/eclipse/jkube/kit/resource/helm/HelmService.java#L75

Sgitario commented on examples/pom.xml:
```diff
@@ -83,5 +83,11 @@
+      <modules>
+        <module>helm-example</module>
+      </modules>
+    </profile>
```
This is because I'm thinking to use HELM in real, but if this is too complex, I will remove it and only verify the artifacts are generated.

Sgitario commented:
Update PR with:
- Use camel case as Helm does not support other name convention
- Add documentation in the helm example module.

aureamunoz reviewed (commented):
it looks great! I did just two minor remarks

Sgitario commented on annotations/s2i-annotations/src/main/java/io/dekorate/s2i/annotation/S2iBuild.java:
```diff
@@ -42,7 +42,7 @@
    * @return The registry or empty string if no registry has been specified.
    */
-  String registry() default "";
+  String registry() default "docker.io";
```
reverted

Sgitario commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/annotation/HelmBuild.java:
```diff
@@ -0,0 +1,61 @@
+@Pojo(name = "HelmBuildConfig", autobox = true, mutable = true, superClass = Configuration.class, relativePath = "../config", withStaticBuilderMethod = true, withStaticAdapterMethod = false, adapter = @Adapter(name = "HelmBuildConfigAdapter", relativePath = "../adapter", withMapAdapterMethod = true))
+@Target({ ElementType.CONSTRUCTOR, ElementType.TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface HelmBuild {
```
Good point! Will change it to `HelmChart`.

aureamunoz reviewed (approved):
LGTM :-)

iocanel commented on core/src/main/java/io/dekorate/kubernetes/decorator/ApplyImageDecorator.java:
```diff
@@ -41,4 +42,19 @@ public void andThenVisit(ContainerFluent container) {
+
+  @Override
+  public String getJsonPathProperty() {
+    return "$.[?(@.kind == 'Deployment')].spec.template.spec.containers[?(@.name == '" + getContainerName() + "')].image";
```
The decorator could be used not only in `Deployment` but in `DeploymentConfig`, knative `Service` and so on.
With this in mind, I think that the json path expression is way too limiting.
Not a json path expert but I think that this might be more appropriate:
`$..containers[?(@.name == '" + getContainerName() +"'].image`

Sgitario replied:
the container name should never be null
Also, the decorator seems to be used only for k8s -> I need to investigate more why.
However, following your comment, I made your requested changes.

iocanel reviewed (changes requested):
I really really like it!

There are some comments regarding the configured json paths that we need to align. Other than that its looks great.

cmoulliard commented:
> ```ini
> dekorate.helm.values[
> ```

I like a lot the approach where we can define using the property file the helm metadata. 
Question: Will it be needed that we define all the values using (value[0],value[1], ...) ? If yes, could we adopt a more user friendly approach where I could define a hierarchy of YAML fields ? @Sgitario

cmoulliard commented on examples/helm-example/readme.md:
````diff
@@ -0,0 +1,53 @@
+Finally, let's use Helm to deploy it into the cluster:
+
+```shell
+helm install helm-example ./target/classes/META-INF/dekorate/
````
As dekorate supports to build/push an image, why don't we add an option to also install/uninstall a chart .... @Sgitario

iocanel replied:
I can see that being useful for making sure that things work, but it doesn't add any additional value, given that the effect is/should be the same as deploying the generated kubernetes manifests.

iocanel commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/annotation/HelmChart.java:
```diff
@@ -0,0 +1,61 @@
+@Pojo(name = "HelmChartConfig", autobox = true, mutable = true, superClass = Configuration.class, relativePath = "../config", withStaticBuilderMethod = true, withStaticAdapterMethod = false, adapter = @Adapter(name = "HelmChartConfigAdapter", relativePath = "../adapter", withMapAdapterMethod = true))
+@Target({ ElementType.CONSTRUCTOR, ElementType.TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface HelmChart {
```
Well, this is a general concern that affects all things generated by dekorate. Not something we should touch in the particular PR, IMHO.

Sgitario commented:
> > ```ini
> > dekorate.helm.values[
> > ```
> 
> I like a lot the approach where we can define using the property file the helm metadata. Question: Will it be needed that we define all the values using (value[0],value[1], ...) ? If yes, could we adopt a more user friendly approach where I could define a hierarchy of YAML fields ? @Sgitario

Users using YAML file to configure the application can use the YAML format. Moreover, users can also define the values using the annotation `@HelmChart`.

cmoulliard commented:
> Users using YAML file to configure the application can use the YAML format.

Can you elaborate please and explain which YAML file we are talking about here ? @Sgitario

Sgitario commented:
> > Users using YAML file to configure the application can use the YAML format.
> 
> Can you elaborate please and explain which YAML file we are talking about here ? @Sgitario

If users are using YAML for their application, they can configure helm via YAML. Example: https://github.com/dekorateio/dekorate/blob/main/tests/feat-216-annotationless-yml/src/main/resources/application.yml

Moreover, users can use the annotation way: https://github.com/dekorateio/dekorate/pull/841/files#diff-2dee07e059e7ee3dbbb449e1ce0190e4e876648ea748e29f9b6261598385ec10R25

cmoulliard commented:
> 

Thanks. What is then the best way  (or our recommended way) to finally define the YAML helm template that next a project will use and will customize using the `values.yaml` file ? @Sgitario

cmoulliard commented:
Question: Will it be able to include some additional METADATA part of the chart generated if by example the framework includes: Health Check,... ? @Sgitario

Sgitario commented:
> > 
> 
> Thanks. What is then the best way (or our recommended way) to finally define the YAML helm template that next a project will use and will customize using the `values.yaml` file ? @Sgitario

We don't need to define a YALM helm template. The framework will create all the helm artifacts based on the user configuration (from annotations and configuration) and the decorators that support the interface `WithConfigReference`.

Sgitario commented:
> Question: Will it be able to include some additional METADATA part of the chart generated if by example the framework includes: Health Check,... ? @Sgitario

I'm speaking with @iocanel at the moment about what other configuration to be included in the chart values.yaml manifest automaically (without user interaction). For example, another important configuration is the s2i builder image.

iocanel reviewed (approved):
LGTM

Sgitario commented:
PR updated to:
- Changed interface from `ConfigReference` to `ConfigReferences`, so one decorator can provide multiple config references
- Changed value reference to provide multiple json paths, so one config reference can replace multiple json paths
- Supported K8s ingress, fyi @aureamunoz

aureamunoz reviewed (approved):
LGTM, great job!

Sgitario commented:
Summary of required changes after review by Anne-Marie:
- Generate Helm artifacts in `classes/META-INF/dekorate/<chart name`
- To follow the Helm standards, generate an empty folder `charts`.
- Split resources into multiple files inteads of having a big `kubernetes.yml` (or `openshift.yml`) file.

iocanel commented:
> * Split resources into multiple files inteads of having a big `kubernetes.yml` (or `openshift.yml`) file.

This is something that may make sense for helm, but I have strong objections to applying this change globally. So let's split them for the charts only.

Sgitario commented:
> * [ ]  Support multi tenant (profiles) for `values.yaml` configuration:
> 
> ```
> dekorate.helm.<profile: in your example `dev`>.values[0].property=helm-example.name
> dekorate.helm.<profile: in your example `dev`>.values[0].jsonPath=$..metadata.name
> 
> And these values will generate the profiled values `values.<profile>.yaml`.
> ```

@iocanel @aureamunoz It seems that we cannot do this because this needs to mimic annotations and annotations do not support maps. As an alternative, I have added a new property `profile` which is available in both the annotation and in the properties:
```
dekorate.helm.values[0].property=helm-example.name
dekorate.helm.values[0].jsonPath=$..metadata.name
dekorate.helm.values[0].profile=dev
```

Are you ok with doing this way?

Sgitario commented:
PR updated with all the suggestions from https://github.com/dekorateio/dekorate/pull/841#issuecomment-1016479954. 
@iocanel @aureamunoz

iocanel reviewed (changes requested):
We should avoid generating arbitrary folders under `classes/META-INF/dekorate` as it might restrict us in the future to generate folders there (see [1]).

So, I would like us to use the helm folder name. So, the helm chart root should be: `classes/META-INF/dekorate/helm/<chart name>`.

[1]: Since, a lot of people prefer the file per resource style, we might consider in the future to provide it as an option. This most likely would be provided by adding an additional folder under dekorate. So, generating folder names there might be a source for clashes.

aureamunoz commented:
> mimic annotations

Ok, and is profile mandatory? is there a default value?

aureamunoz commented:
> 

Ok, and is profile mandatory? is there a default value?

Sgitario commented:
> > 
> 
> Ok, and is profile mandatory? is there a default value?

No, it's an optional value. If not provided, the config will be part of the `values.yaml` (which is meant for default profile)

aureamunoz commented:
> > > 
> > 
> > 
> > Ok, and is profile mandatory? is there a default value?
> 
> No, it's an optional value. If not provided, the config will be part of the `values.yaml` (which is meant for default profile)

I've just seen the code :-)

Sgitario commented:
> We should avoid generating arbitrary folders under `classes/META-INF/dekorate` as it might restrict us in the future to generate folders there (see [1]).
> 
> So, I would like us to use the helm folder name. So, the helm chart root should be: `classes/META-INF/dekorate/helm/<chart name>`.
> 
> [1]: Since, a lot of people prefer the file per resource style, we might consider in the future to provide it as an option. This most likely would be provided by adding an additional folder under dekorate. So, generating folder names there might be a source for clashes.

Good point! I've updated the PR with this change already.

cmoulliard commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/annotation/HelmChart.java:
```diff
@@ -0,0 +1,61 @@
+@Pojo(name = "HelmChartConfig", autobox = true, mutable = true, superClass = Configuration.class, relativePath = "../config", withStaticBuilderMethod = true, withStaticAdapterMethod = false, adapter = @Adapter(name = "HelmChartConfigAdapter", relativePath = "../adapter", withMapAdapterMethod = true))
+@Target({ ElementType.CONSTRUCTOR, ElementType.TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface HelmChart {
```
Is this HelmChart definition still in line with Helm v3 Chart Go struct (see https://github.com/helm/helm/blob/v3.7.2/pkg/chart/chart.go, https://github.com/helm/helm/blob/v3.7.2/pkg/chart/metadata.go,https://github.com/helm/helm/blob/v3.7.2/pkg/chart/dependency.go) ?

Sgitario replied:
Yes, though all the fields are not mapped. I've only mapped the same ones as in JKube.

cmoulliard commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/manifest/HelmManifestGenerator.java:
```diff
@@ -0,0 +1,65 @@
+  public void generate(HelmChartConfig config) {
+    if (config.isEnabled()) {
+      LOGGER.info("Processing helm configuration.");
+      // TODO
```
Implement the `TODO` or remove the comment `//TODO` ;-)

Sgitario replied:
Well spotted! Actually, the `HelmManifestGenerator` is not needed at all. Removed.

cmoulliard commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/model/Chart.java:
```diff
@@ -0,0 +1,154 @@
+
+/**
+ * Represents the <a href="https://github.com/kubernetes/helm">Helm</a>
+ * <a href="https://github.com/kubernetes/helm/blob/master/pkg/proto/hapi/chart/metadata.pb.go#L50">Chart.yaml file</a>
```
Link is not correct anymore and should be updated for all href mentioned within the classes !!

Sgitario replied:
Well spotted! Comment updated.

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,43 @@
+
+How can we use it?
+
+First, make sure you have logged into a cluster with Helm enabled (in OpenShift clusters, Helm is enabled by default).
```
This sentence is not correct. Since helm v3, no server component is deployed anymore on k8s or openshift. So, currently it is only needed that helm v3 is installed locally.

Sgitario replied:
Updated to:
```
First, make sure you have installed [the Helm command line](https://helm.sh/docs/intro/install/) and have connected into a cluster.
```

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,43 @@
+
+First, make sure you have logged into a cluster with Helm enabled (in OpenShift clusters, Helm is enabled by default).
+
+Then, generate the Helm artifacts and push the image into a container registry:
```
Can you maybe mention that the following line will not only generate the Helm chart for the project but also build/push the container runtime image to a registry

Sgitario replied:
Updated with:
```
Then, run the following Maven command in order to generate the Helm artifacts and build/push the image into a container registry:
```

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
````diff
@@ -0,0 +1,43 @@
+Finally, let's use Helm to deploy it into the cluster:
+
+```shell
+helm install helm-example ./target/classes/META-INF/dekorate/helm/<chart name>
````
I propose to include also 1-2 additional examples to show how to override the values.yaml using `--set` or where you pass `my-values.yaml`

Sgitario replied:
Added

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
````diff
@@ -0,0 +1,43 @@
+helm install helm-example ./target/classes/META-INF/dekorate/helm/<chart name>
+```
+
+Or upgrade your Helm deployment:
````
Command to uninstall the chart is missing ;-)

Sgitario replied:
Added

cmoulliard commented on examples/helm-on-kubernetes-example/readme.md:
```diff
@@ -0,0 +1,53 @@
+# Helm Example 
```
This `Helm example` content should be aligned with `NOTES.txt` content

Sgitario replied:
Done

cmoulliard commented on examples/helm-on-kubernetes-example/src/main/resources/application.properties:
```diff
@@ -0,0 +1,22 @@
+dekorate.helm.name=myChart
+# Normal use case
+dekorate.helm.values[0].property=helmOnKubernetesExample.name
+dekorate.helm.values[0].jsonPaths=$..metadata.name,$.[?(@.kind == 'Ingress')].spec.rules..http.paths..backend.service.name
```
The syntax is very complex. Can we document it please and perhaps add a doc link where developers could discover how to design such JSONpath query ?

Sgitario replied:
Added javadoc to ValueReference.jsonPaths with links about how to use it.
For the rest of the document, it will be done in https://github.com/dekorateio/dekorate/issues/842 once the Dekorate site is merged.

Sgitario commented:
@cmoulliard I've already done all your requested changes.

cmoulliard commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/model/Chart.java:
```diff
@@ -25,8 +25,8 @@
- * Represents the <a href="https://github.com/kubernetes/helm">Helm</a>
- * <a href="https://github.com/kubernetes/helm/blob/master/pkg/proto/hapi/chart/metadata.pb.go#L50">Chart.yaml file</a>
+ * Represents the <a href="https://github.com/helm/helm">Helm</a>
+ * <a href="https://github.com/helm/helm/blob/c7e1f9b04606574b2f0d93d34e22efb06847ad08/pkg/chart/metadata.go#L44">Chart.yaml file</a>
```
Why do yu use a commit instead of a tagged version (e.g. https://github.com/helm/helm/blob/v3.7.2/pkg/chart/metadata.go) ?

Sgitario replied:
The link was generated by Github. Changed.

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -20,9 +20,9 @@ You can find the generated Helm artifacts under: `target/classes/META-INF/dekora
 How can we use it?
 
-First, make sure you have logged into a cluster with Helm enabled (in OpenShift clusters, Helm is enabled by default).
+First, make sure you have installed [the Helm command line](https://helm.sh/docs/intro/install/) and have connected into a cluster.
```
I suggest to say instead of `and have connected into a cluster.` --> `and has access to a kubernetes cluster`

cmoulliard commented on examples/helm-on-kubernetes-example/readme.md:
```diff
@@ -22,9 +22,9 @@ You can find the generated Helm artifacts under: `target/classes/META-INF/dekora
 How can we use it?
 
-First, make sure you have logged into a cluster with Helm enabled (in OpenShift clusters, Helm is enabled by default).
+First, make sure you have installed [the Helm command line](https://helm.sh/docs/intro/install/) and have connected into a cluster.
```
See previous comment

Sgitario replied:
Changed

cmoulliard commented on examples/helm-on-kubernetes-example/readme.md:
````diff
@@ -38,16 +38,36 @@ Finally, let's use Helm to deploy it into the cluster:
 helm install helm-example ./target/classes/META-INF/dekorate/helm/<chart name>
 ```
 
+The above command will use the default values file `values.yaml`, we can select a different values file by doing:
````
This is not clear. I suggest to say "To override the default values, part of the `values.yaml` file, pass as parameter you own value file `-f /path/to/the/file/values.dev.yaml` or set them using `--set key=val --set key=val ...`"

Sgitario replied:
Changed

geoand commented on examples/helm-on-kubernetes-example/readme.md:
```diff
@@ -0,0 +1,73 @@
+
+- Via the `upgrade` option of Helm command line:
+
+After doing changes in your project, you would need to regenerate the resources using Dekorate:
```
```suggestion
After making changes to your project, you would need to regenerate the resources using Dekorate:
```

Sgitario replied:
Done

geoand commented on examples/helm-on-openshift-example/readme.md:
```diff
@@ -0,0 +1,77 @@
+
+- Via the `upgrade` option of Helm command line:
+
+After doing changes in your project, you would need to regenerate the resources using Dekorate:
```
Same here

geoand commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,71 @@
+      <version>${project.version}</version>
+    </dependency>
+
+Compile the project using:
```
```suggestion
Build the project using:
```

Sgitario replied:
Changed

geoand commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,71 @@
+- values.yaml
+- templates/*.yml the generated resources by Dekorate
+
+How can we use it?
```
```suggestion
How can it be used? 
```

Sgitario replied:
Changed

geoand commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
````diff
@@ -0,0 +1,71 @@
+helm install helm-example ./target/classes/META-INF/dekorate/helm/<chart name> --values ./target/classes/META-INF/dekorate/helm/<chart name>/values.dev.yaml
+```
+
+How can we update my deployment?
````
```suggestion
How can I update my deployment?
```

Sgitario replied:
Changed

geoand commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,71 @@
+
+- Via the `upgrade` option of Helm command line:
+
+After doing changes in your project, you would need to regenerate the resources using Dekorate:
```
Same comment as other files

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+Dekorate also supports generating manifests for `helm`. To make use of this feature you need to add:
+
+```xml
+<dependency>
````
General remark for the Dekorate doc. If a gradle user is interested to use Dekorate, is there a section mentioning what they should do to configure also the gradle file managing the dependencies !!

Sgitario replied:
+100 
I created this issue to document Dekorate using Gradle: https://github.com/dekorateio/dekorate/issues/849

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+</dependency>
+```
+
+This module provides the [@HelmChart](annotations/helm-annotations/src/main/java/io/dekorate/helm/annotation/HelmChart.java) that will generate the following Helm resources:
````
Is it the proper wording to be used. Do we name all the features (Kubernetes, Docker, Knative, ...) of Dekorate as `modules` ?

Remark: I think that we should rephrase the beginning of this chapter and start with:
- First introduce helm dekorate
- Next, explain what you must do to generate the Helm manifests
- ...

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,277 @@
+
+#### Getting Started
+
+Add your Helm chart configuration in your properties like:
```
Please better introduce this getting started section as we cannot start with `Add your Helm chart ...`

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,277 @@
+
+#### Building
+
+    mvn clean package
+
+To generate the above Helm resources under the folder `target/classes/META-INF/dekorate/helm/<chart name>/`.
```
Should be reviewed please !

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,277 @@
+
+To generate the above Helm resources under the folder `target/classes/META-INF/dekorate/helm/<chart name>/`.
+
+Moreover, assuming you're using Kubernetes, the Helm templates will include the following files by default:
```
Add the link to the kubernetes dekorate section and also mention that you can also generate the resources fro openshift + link

Sgitario replied:
Done

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+  replicas: 3
+```
+
+Note that `myModule` is the name of your project.
````
Highlight the Note with something like `**Note**: `

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,277 @@
+
+##### Mapping user properties using JSONPath expressions
+
+What about if you want to map another property like the name of all the resources (the metadata name)?
```
To be clarify and give a real or better example. I understand your intention but I dont think that the readers will understand. So the idea here is basically to set the name, namespace or labels, ... of the Kubernetes resources generated without the need to find/replace all of them part f the different files ...

Sgitario replied:
Rephrased. Let me know if it looks clearer now.

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+What about if you want to map another property like the name of all the resources (the metadata name)?
+Dekorate allows users to define [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html) expressions to map properties into the Helm values file.
+
+For example, having the following YAML resource:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: helm-on-kubernetes-example
+...
+```
+
+After getting familiar with JSONPath, the expression to map the metadata name value is `$..metadata.name`, so you need to add it to your configuration:
````
You must first explain what you would like to do and not saying `For example ...`. So the phrase should start as:
"If you would like to change the metadata/name of the following YAML file resource ... then you will have to declare part of the property file JSONPath` expression such as `` ...."

Sgitario replied:
Rephrased. Let me know if it looks clearer now.

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+dekorate.helm.values[0].jsonPaths=$..metadata.name
+```
+
+| Note that JSONPath works with JSON, not YAML. So, for testing your expressions, you first need to convert YAML files to JSON files using [this online tool](https://jsonformatter.org/yaml-to-json) and then use some JSONPath online evaluator like [https://jsonpath.com/](https://jsonpath.com/). Remember to wrap the json objects into lists.
````
The Note should be moved after the example presentation !

Sgitario replied:
Moved it up. Let me know if it's clearer now.

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+  name: helm-on-kubernetes-example
+```
+
+Why the value is `helm-on-kubernetes-example`? This is because Dekorate will automatically inspect the generated resources to find the matching value of the JSONPath expression. 
````
Should be rephrased --> `Why is the value generated equal to "" ?`

Sorry but I dont understand the response  as I dont see the link between what it is generated as new `metadata/name` and what you reply "This is because Dekorate will automatically inspect the generated resources to find the matching value of the JSONPath expression".

Sgitario replied:
+1 this might result confusing. I've repharsed this sentece to be more direct.

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,277 @@
+
+##### Mapping multiple properties at once
+
+What about if the properties to map are at different locations, for example:
```
Rephrase to `are located in different places` AND better detail what is the purpose of this example. What do you want to find and replace ...

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,277 @@
+  host: my-test-host
+```
+
+#### Helm Usage
````
Align this text with previous remark :up:

Sgitario replied:
What do you mean? I think it's ok as it is. The sections are:
- Getting Started
- Building
- Mapping Values
- Helm Profiles
- Helm Usage

All are at the same level.

Sgitario commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/annotation/HelmChart.java:
```diff
@@ -0,0 +1,59 @@
+import io.sundr.builder.annotations.Adapter;
+import io.sundr.builder.annotations.Buildable;
+import io.sundr.builder.annotations.Pojo;
+
```
Good point. Added.

Sgitario commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,71 @@
+
+    mvn clean install
+
+You can find the generated Helm artifacts under: `target/classes/META-INF/dekorate/helm/<chart name>/` that should look like:
```
Added

Sgitario commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
````diff
@@ -0,0 +1,71 @@
+mvn clean package -Ddekorate.push=true -Ddekorate.docker.registry=<container registry url> -Ddekorate.docker.group=<your group>
+```
+
+This command will push the image into the container registry to be available for the cluster when deploying.
````
Changed

Sgitario commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
````diff
@@ -0,0 +1,71 @@
+helm install helm-example ./target/classes/META-INF/dekorate/helm/<chart name>
+```
+
+The above command will use the default values file `values.yaml`, we can select a different values file by doing:
````
The proposed sentence is not clear to me. 
I changed it to `The above command will deploy a chart using the default values (as defined within the `values.yaml` file). We can override the default values to use your `values.dev.yaml` file by doing:`. 
Let me know if it's ok to you.

Sgitario commented on assets/config.md:
```diff
@@ -366,6 +367,47 @@ The section below describes all the available subtypes.
 | dekorate.knative.revision-auto-scaling           | AutoScaling                   | Revision autoscaling configuration.                                                                                                                                                                                                                                                                                        ...
 | dekorate.knative.global-auto-scaling             | GlobalAutoScaling             | Global autoscaling configuration.                                                                                                                                                                                                                                                                                          ...
 
+## Helm
```
We need to maintain it statically :(

Sgitario commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,277 @@
+
+### Helm
+
+Dekorate also supports generating manifests for `helm`. To make use of this feature you need to add:
```
changed

Sgitario commented:
@cmoulliard I've already updated the PR with your requested changes.

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
````diff
@@ -0,0 +1,73 @@
+helm install helm-example ./target/classes/META-INF/dekorate/helm/<chart name>
+```
+
+The above command will deploy a chart using the default values (as defined within the `values.yaml` file). We can override the default values to use your `values.dev.yaml` file by doing:
````
by doing --> by executing the following command:

Sgitario replied:
Changed

cmoulliard commented on annotations/helm-annotations/src/main/resources/NOTES.txt:
```diff
@@ -0,0 +1,73 @@
+
+How can it be used?
+
+First, make sure you have installed [the Helm command line](https://helm.sh/docs/intro/install/) and have connected into a cluster.
```
"and have connected into a cluster." --> "and connected/logged to a kubernetes cluster."

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,276 @@
+
+#### Getting Started
+
+To generate the Helm resources, you first need to configure the Helm chart via properties:
```
Is it mandatory to perform that in order to generate the helm resources ? I mean to create 2 properties part of the `properties.txt` file ? If yes, why don t you include the app version, ... as they exist with a Chart yaml file that we can have using `helm init` ?

Sgitario replied:
It's not mandatory. However, I see your point, will update this.

Sgitario replied:
Changed

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,276 @@
+}
+```
+
+Once, you have configured the Helm chart, you can generate the Helm resources under the folder `target/classes/META-INF/dekorate/helm/<chart name>/` with:
````
"with:" --> "using themaven command: "

Sgitario replied:
changed

cmoulliard commented on docs/documentation/helm.md:
```diff
@@ -0,0 +1,276 @@
+
+    mvn clean package
+
+Assuming you're using the [Kubernetes]({{site.baseurl}}/docs/kubernetes) Dekorate extension, the Helm resources will include by default the following templates at `target/classes/META-INF/dekorate/helm/<chart name>/templates/`:
```
If your project declared the [Kubernetes]({{site.baseurl}}/docs/kubernetes) Dekorate extension, then the helm resources the following templates at `target/classes/META-INF/dekorate/helm/<chart name>/templates/`:"

What will be generated if now the project do not include kubernetes or openshift extension ?

Sgitario replied:
I don't think we should include all the possible templates that are generated depending on the extension in use. 
However, I rephrased this sentence again, let me know if now it's clearer.

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,276 @@
+...
+```
+
+Dekorate will not replace the property at `metadata.name` with `{{ .Values.myModule.name }}`.
````
There is still something which is missing as you dont do a transition within the text between "Let's see the following YAML resource:" AND "Dekorate will not replace the property at `metadata.name` with `{{ .Values.myModule.name }}`.".

Please add a line doing such a transition as we still dont know yet what you would like to do

Sgitario replied:
Done

cmoulliard commented on docs/documentation/helm.md:
````diff
@@ -0,0 +1,276 @@
+            pathType: Prefix
+```
+
+For this scenario, you need to provide a comma-separated list of JSONPath expressions to be mapped to the same property `myModule.name`:
````
Same remark as before. Transition text is still missing as the user still don't know what you would like to do here.

Sgitario replied:
I see what you meant now. Updated.

Sgitario commented:
@cmoulliard PR updated again. Thanks!

iocanel reviewed (approved):
I have added some minor comments / questions.
As, they are mostly cleanup related they could be possibly resolved in followup PRs too.
I will approve and leave it to @Sgitario to decide how to deal with it.

Sgitario commented on annotations/helm-annotations/pom.xml:
```diff
@@ -0,0 +1,181 @@
+                <filter>
+                  <artifact>*:*</artifact>
+                  <includes>
+                    <include>io/dekorate/s2i/adapter/**</include>
```
fixed

Sgitario commented on annotations/helm-annotations/src/main/java/io/dekorate/helm/config/DefaultHelmBuildConfigGenerator.java:
```diff
@@ -0,0 +1,40 @@
+
+  public DefaultHelmBuildConfigGenerator(ConfigurationRegistry configurationRegistry) {
+    this.configurationRegistry = configurationRegistry;
+    this.configurationRegistry.add(new ApplyProjectInfo(getProject()));
```
Good point!

Sgitario commented on annotations/helm-annotations/pom.xml:
```diff
@@ -0,0 +1,181 @@
+
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>openshift-client</artifactId>
```
Copy-paste from other modules. Removed.

Sgitario commented:
> I have added some minor comments / questions. As, they are mostly cleanup related they could be possibly resolved in followup PRs too. I will approve and leave it to @Sgitario to decide how to deal with it.

I've already done all the requested changes. Many thanks @iocanel !
